core/message_queue_test.go

```diff
@@ -0,0 +1,291 @@
+		assertNoNonce(t, q, alice)
+
+		mustEnqueue(t, q, msgs[0], 0)
+		mustEnqueue(t, q, msgs[1], 0)
+		assertLargestNonce(t, q, alice, 1)
+	})
+
+	t.Run("independent addresses", func(t *testing.T) {
+		fromAlice := []*types.SignedMessage{
+			newMessage(t, alice, 0),
+			newMessage(t, alice, 1),
+			newMessage(t, alice, 2),
+		}
+		fromBob := []*types.SignedMessage{
+			newMessage(t, bob, 10),
+			newMessage(t, bob, 11),
+			newMessage(t, bob, 12),
+		}
+		q := core.NewMessageQueue()
+
+		mustEnqueue(t, q, fromAlice[0], 0)
+		assertNoNonce(t, q, bob)
+
+		mustEnqueue(t, q, fromBob[0], 0)
+		assertLargestNonce(t, q, alice, 0)
+		assertLargestNonce(t, q, bob, 10)
+
+		mustEnqueue(t, q, fromBob[1], 0)
+		mustEnqueue(t, q, fromBob[2], 0)
+		assertLargestNonce(t, q, bob, 12)
+
+		mustEnqueue(t, q, fromAlice[1], 0)
+		mustEnqueue(t, q, fromAlice[2], 0)
+		assertLargestNonce(t, q, alice, 2)
+
+		msg := mustRemoveNext(t, q, alice, 0)
+		assert.Equal(fromAlice[0], msg)
+
+		msg = mustRemoveNext(t, q, bob, 10)
+		assert.Equal(fromBob[0], msg)
+
+		q.Clear(bob)
+		assertLargestNonce(t, q, alice, 2)
+		assertNoNonce(t, q, bob)
+	})
+
+	t.Run("expire before stamp", func(t *testing.T) {
+		fromAlice := []*types.SignedMessage{
+			newMessage(t, alice, 0),
+			newMessage(t, alice, 1),
+		}
+		fromBob := []*types.SignedMessage{
+			newMessage(t, bob, 10),
+			newMessage(t, bob, 11),
+		}
+		q := core.NewMessageQueue()
+
+		mustEnqueue(t, q, fromAlice[0], 100)
+		mustEnqueue(t, q, fromAlice[1], 101)
+		mustEnqueue(t, q, fromBob[0], 200)
+		mustEnqueue(t, q, fromBob[1], 201)
+
+		assert.Equal(uint64(100), q.NextStamp(alice))
+		assert.Equal(uint64(200), q.NextStamp(bob))
+
+		expired := q.ExpireBefore(0)
+		assert.Empty(expired)
+
+		expired = q.ExpireBefore(100)
+		assert.Empty(expired)
+
+		// Alice's whole queue expires as soon as the first one does
+		expired = q.ExpireBefore(101)
+		assert.Equal(map[address.Address][]*types.SignedMessage{
+			alice: {fromAlice[0], fromAlice[1]},
+		}, expired)
+
+		assert.Equal(uint64(0), q.NextStamp(alice))
+		assertNoNonce(t, q, alice)
+		assert.Equal(uint64(200), q.NextStamp(bob))
+		assertLargestNonce(t, q, bob, 11)
+
+		expired = q.ExpireBefore(300)
+		assert.Equal(map[address.Address][]*types.SignedMessage{
+			bob: {fromBob[0], fromBob[1]},
+		}, expired)
+
+		assert.Equal(uint64(0), q.NextStamp(bob))
+		assertNoNonce(t, q, bob)
+	})
+}
+
+// Creates a new message from the ith address with a specified nonce
+func newMessage(t *testing.T, from address.Address, nonce uint64) *types.SignedMessage {
+	seq := msgSeq
+	msgSeq++
+	msg := types.NewMessage(
+		from,
+		address.NewMainnet([]byte("destination")),
+		nonce,
+		types.NewAttoFILFromFIL(0),
+		"method"+fmt.Sprintf("%d", seq),
+		[]byte("params"))
+	signed, err := types.NewSignedMessage(*msg, signer, gasPrice, gasUnits)
+	require.NoError(t, err)
+	return signed
+}
+
+func mustEnqueue(t *testing.T, q *core.MessageQueue, msg *types.SignedMessage, stamp uint64) {
```

Comment:
We should standardize on naming for these things that call require.NoError and friends.  I don't have any preference, but I've seen both mustDoAThing and requireDoAThing.  I'll even go back and change my own stuff if people prefer mustDoAThing.

Reply:
I think I saw `must` pattern in godocs somewhere, but can't find it now.

Reply:
I'll resolve in a follow-up if I've ended up doing the wrong thing (seems likely)